Date conversion - first solution

rositsabodurova opened the pull request:
@NewestUser 
@IvoStaynov 
@dahmedov 
Here is what I have figured out as a first simple solution for the date conversion. I will be expecting your ideas and comments.

rositsabodurova committed:
first try

NewestUser commented on src/main/java/com/newestuser/imen/ImenApplication.java:
```diff
@@ -1,13 +1,19 @@
 import org.springframework.boot.SpringApplication;
 import org.springframework.boot.autoconfigure.SpringBootApplication;
 
+import java.util.Date;
```
please remove redundant imports and white space.

Imports can be cleared by intellij with **ctrl + alt + o** or **ctrl + shift + o** (i am not sure how it is in WIndows) The option is called *Optimize imports* you can google it

rositsabodurova replied:
Ctrl + alt + o it is. It removes unnecessary imports and rearranges the others.

NewestUser commented on src/main/java/com/newestuser/imen/web/BulgarianDate.java:
```diff
@@ -0,0 +1,299 @@
+    private static String findYearType(int bulgarianDateYear) {
+        int bulgarianDateYearTypeIndex = bulgarianDateYear % 12;
+        String bulgarianDateYearType = "";
+        switch (bulgarianDateYearTypeIndex) {
```
This switch/case can be implemented using a `Map`.

Example:

```java
    private static Map<Integer, String> YEAR_TYPES = new HashMap<>() {{
        put(0, "Верени");
        put(1, "Дилом");
        put(2, "Имен");
        put(3, "Теку");
        put(4, "Бечин");
        put(5, "Тох");
        put(6, "Етх");
        put(7, "Докс");
        put(8, "Сомор");
        put(9, "Шегор");
        put(10, "Барс");
        put(11, "Дван");
    }};

    private static String findYearType(int bulgarianDateYear) {

        return YEAR_TYPES.get(bulgarianDateYear % 12);
    }

```

NewestUser commented on src/main/java/com/newestuser/imen/web/BulgarianDate.java:
```diff
@@ -0,0 +1,299 @@
+        return bulgarianDateYearType;
+    }
+
+    public static String returnBulgarianDate() {
```
I understand this is your representation of the current time according to the BG Calendar.

Most date/time libraries that I have used use the word `now` to represent the current instance. We can borrow that convention.

Here is how it would look:
```java
String now = BulgarianDate.now();
```

NewestUser commented on src/main/java/com/newestuser/imen/web/IndexController.java:
```diff
@@ -13,10 +13,7 @@
-        Date currentDate = new Date();
-
-        model.addAttribute("name", dateFormat.format(currentDate));
+        model.addAttribute("name", BulgarianDate.returnBulgarianDate());
```
I would suggest that you rename the parameter `name` to something more meaningful like `currentTime` or `now` or `date` or something similar. This would include a change in the HTML template.

This was committed by me by mistake.

NewestUser commented on src/main/java/com/newestuser/imen/web/BulgarianDate.java:
```diff
@@ -0,0 +1,299 @@
+import java.util.Calendar;
+
+
+public class BulgarianDate {
```
I would recommend you write at-least one unit test. This would make refactoring much easier especially if it needs to be done by other people that are not that familiar with the code.

I am not insisting on more unit tests because it might be too much for a first PR :)

In order to write a unit test open this class in IntelliJ and press CTRL + ALT + T. IntelliJ will prompt you to create a test class. Select that option and press enter.

Make sure this is the configuration that you choose:
<img width="568" alt="Screenshot 2019-10-06 at 13 40 24" src="https://user-images.githubusercontent.com/7418570/66267825-e76ff000-e83e-11e9-9d2d-7fce97e3ce16.png">


Here is how one test method might look like in the generated by IntelliJ class:

```java
    @Test
    public void formatGregorianDateToBulgarianDate() {
        Date gregorianDate = firstOfJanuary(2019);

        BulgarianDate bulgarianDate =  new BulgarianDate(gregorianDate);
        String formattedDate = bulgarianDate.format();

        assertEquals(formattedDate,"10 Алем, 6787 Докс" );
    }

    private Date firstOfJanuary(int year){
        Calendar calendar = Calendar.getInstance();
        calendar.set(Calendar.YEAR, year);
        calendar.set(Calendar.MONTH, Calendar.JANUARY);
        calendar.set(Calendar.DAY_OF_MONTH, 1);
        return calendar.getTime();
    }
```
The code above is not compiling. Several changes need to be made. 
1. pass the date gregorian date in the constructor this would make BulgarianDate more flexible and unit testable.
2. Separate formatting the date from calculating the date (see that formatting is performed in a method called `format`

NewestUser commented on src/main/resources/templates/index.html:
```diff
@@ -43,7 +43,7 @@
     <div class="round-border">
         <h1><span th:text="${name}"/></h1>
-        <span class="secondaryText" th:text="${name}"></span>
+        <span class="secondaryText" th:text="${currentDate}"></span>
```
you do not pass a parameter called currentDate in the Controller.
You can pass both the Bulgarian and the Gregorian dates to be visualised.
 
Example:
```java
SimpleDateFormat dateFormat = new SimpleDateFormat("EEEE, MMMM d, yyyy");
        
        Date gregorianDate = new Date();
        BulgarianDate bulgarianDate = BulgarianDate.now();

        model.addAttribute("bulgarianDate", bulgarianDate.format());
        model.addAttribute("gregorianDate", dateFormat.format(gregorianDate));

```

NewestUser commented on src/main/java/com/newestuser/imen/web/BulgarianDate.java:
```diff
@@ -0,0 +1,299 @@
+    private static String convertMonth(int month, int currentDay) {
+        String bulgarianMonth = "";
+        switch (month) {
+            case 1:
```
we can add another abstraction to simplify this long switch case statement.
We can have a class that represents a given day range and it returns the month based on a day. Example:

```java
/**
 * This class is responsible for determining the name of the month based on the gregorian day.
 * Depending on which gregorian day the bulgarian month might vary.
 * Example: from 22of December up until 22nd of January it is the bulgarian month 'Алем'.
 */
public class DayMonthRange  {
    private final int startDay;
    private final int endDay;
    private final String monthInRange;
    private final String monthOutOfRange;

    public DayMonthRange(int startDay, int endDay, String monthInRange, String monthOutOfRange) {
        this.startDay = startDay;
        this.endDay = endDay;
        this.monthInRange = monthInRange;
        this.monthOutOfRange = monthOutOfRange;
    }

    /**
     * Check if the day is in the configured range. The matching is inclusive.
     *
     * @param dayOfMonth the day in a given gregorian month.
     * @return the formatted bulgarian month.
     */
    public String getBulgarianMonth(int dayOfMonth) {
        if (startDay >= dayOfMonth && endDay <= dayOfMonth) {
            return monthInRange;
        }
        return monthOutOfRange;
    }
}

class BulgarianDate {
    private static final Map<Integer, DayMonthRange> GREGORIAN_MONTH_TO_BG_MONTH = new HashMap<>(){{
            put(1, new DayMonthRange(1, 22, "Алем", "Тутом"));
            put(2, new DayMonthRange(1, 21, "Тутом", "Читем"));
            put(3, new DayMonthRange(1, 22, "Читем", "Твирем"));
            put(4, new DayMonthRange(1, 22, "Твирем", "Вечем"));
            put(5, new DayMonthRange(1, 22, "Вечем", "Шехтем"));
            put(6, new DayMonthRange(1, 21, "Шехтем", "Сетем"));
            put(7, new DayMonthRange(1, 22, "Сетем", "Бехти"));
            put(8, new DayMonthRange(1, 22, "Бехти", "Нунтем"));
            put(9, new DayMonthRange(1, 20, "Нунтем", "Елем"));
            put(10, new DayMonthRange(1, 21, "Елем", "Ениалем"));
            put(11, new DayMonthRange(1, 20, "Аниалем", "Алтом"));
            put(12, new DayMonthRange(1, 20, "Алтом", "Алем"));
        }};


  private static String convertMonth(int month, int currentDay) {
        if(month == 12 && currentDay == 21){
            return "Нова година";
        }

        DayMonthRange dayMonthRange = GREGORIAN_MONTH_TO_BG_MONTH.get(month);

        return dayMonthRange.getBulgarianMonth(currentDay);
}


}
```

rositsabodurova replied:
@NewestUser 
The logic in the if-else statement of the getBulgarianMonth() method should be the opposite.

NewestUser commented on src/main/java/com/newestuser/imen/web/BulgarianDate.java:
```diff
@@ -0,0 +1,299 @@
+package com.newestuser.imen.web;
```
I wouldn't place this logic in the `web` package. Calculating the dates has nothing to do with the HTTP web layer. 

I would either place it in the `imen` package or better yet create a separate package for this, something like `time`, `date`, `bgtime`, `bgdate` or something in that spirit.

rositsabodurova committed:
implementing changes

NewestUser commented on src/main/java/com/newestuser/imen/ImenApplication.java:
```diff
@@ -1,13 +1,19 @@
 	public static void main(String[] args) {
 		SpringApplication.run(ImenApplication.class, args);
+
+
```
still a lot of empty line :)

rositsabodurova committed:
implemented conversion methods using maps

rositsabodurova commented:
Hi everyone!
I have addressed most of the comments provided by @NewestUser Thank you for the suggestions and ideas. The unit test is to be done next.

The conversion methods are implemented using Maps. Let me know if you have more elegant ideas for them (especially the one for day conversion). 

I have also added an element to show the Gregorian current date to the index page along with the Bulgarian one.

NewestUser commented on src/main/java/com/newestuser/imen/web/IndexController.java:
```diff
@@ -1,5 +1,7 @@
 package com.newestuser.imen.web;
 
+//import com.newestuser.imen.bgdate.BulgarianDate;
```
unused import

NewestUser commented on src/main/java/com/newestuser/imen/bgdate/DayMonthRange.java:
```diff
@@ -0,0 +1,49 @@
+     * @param currentDayOfMonth the day in the gregorian month.
+     * @return true if the given day is in the first part of the gregorian month.
+     */
+
```
this comment is floating in mid air, 

The docs of a method should be attached to the method

rositsabodurova committed:
Unit tests implementatation attempt

rositsabodurova commented:
Hi everyone!
I need your input here. As you can see from my previous commit, BulgarianDate class does not have methods/constructors for conversion of a given Gregorian date to the respective Bulgarian Date. The conversion happens in the .now() method and affects only the current date. 
I have added a constructor with Date type parameter for Gregorian date, along with an empty one. 
On the other hand, the tests must be performed with specific dates, not only the current one.

The test I have written does not pass, so I need your ideas on that.
@NewestUser 
@dahmedov 
@IvoStaynov 
@b0b3t0 

Thanks!

rositsabodurova committed:
Modifications with working test

NewestUser commented on src/main/java/com/newestuser/imen/bgdate/BulgarianDate.java:
```diff
@@ -0,0 +1,174 @@
+        this.bulgarianDateMonth = convertMonth(calendar.get(Calendar.MONTH) + 1, calendar.get(Calendar.DAY_OF_MONTH));
+        this.bulgarianDateYear = calendar.get(Calendar.YEAR)+ 4768;
+        this.bulgarianDateTypeOfYear = findYearType(this.bulgarianDateYear);
+//        int dayOfWeek = calendar.get(Calendar.DAY_OF_WEEK);
```
remove redundant comments

NewestUser commented on src/main/java/com/newestuser/imen/bgdate/BulgarianDate.java:
```diff
@@ -0,0 +1,174 @@
+
+
+    public BulgarianDate() {
+
```
you can chain the constructors so that the default constructor instantiates a BulgarianDate for the current day

```java
public BulgarianDate() {
   this(new Date());
}
```

Currently some other part of the code can use this constructor to instantiate a date that would have illegal state.

NewestUser commented on src/main/java/com/newestuser/imen/bgdate/BulgarianDate.java:
```diff
@@ -0,0 +1,174 @@
+     * @param currentDay current day in a given gregorian month.
+     * @return the converted Bulgarian month.
+     */
+
```
make sure that you do not have floating documentation here and on the other methods as well

NewestUser commented on src/main/java/com/newestuser/imen/bgdate/BulgarianDate.java:
```diff
@@ -0,0 +1,174 @@
+
+    public static BulgarianDate now() {
+        BulgarianDate bulgarianDate = new BulgarianDate();
+        bulgarianDate.bulgarianDateDay = findBulgarianDay(Calendar.getInstance().get(Calendar.DAY_OF_MONTH), bulgarianDate.bulgarianDateMonth, Calendar.getInstance().get(Calendar.MONTH) + 1);
```
I am not sure why you pass `bulgarianDtae.buglarianDateMonth` as a parameter to `findBulgarianDay` when you attempt to calculate the bulgarianDay when that parameter is not even initialized.

You can remove that parameter from `findBulgarianDay` you do not use it.

NewestUser replied:
Overall all this entire logic is duplicated in the constructor.

If you chain the constructors this method can become:

```java
public static BulgarianDate now() {
   return new BulgarianDate();
}
```

And you would be able to remove the `static` modified from all the private methods.

NewestUser commented on src/main/java/com/newestuser/imen/bgdate/BulgarianDate.java:
```diff
@@ -0,0 +1,174 @@
+
+public class BulgarianDate {
+    private int bulgarianDateDay;
+    private String bulgarianDateMonth;
```
you can make all the fields final

NewestUser commented on src/main/java/com/newestuser/imen/bgdate/DayMonthRange.java:
```diff
@@ -0,0 +1,49 @@
+     * @param dayOfMonth the day in a given gregorian month.
+     * @return the formatted Bulgarian month.
+     */
+
```
floating doc

rositsabodurova committed:
comments and modifications implemented

NewestUser commented:
LGTM

Your build is failing:
`Failed to execute goal org.apache.maven.plugins:maven-pmd-plugin:3.12.0:check (validate) on project imen: You have 2 PMD violations.`

You can run `./mvnw clean verify` to see the results.
You need to fix them as this blocks merging this PR

rositsabodurova committed:
adding missing features and updates

rositsabodurova committed:
adding missing features and updates

rositsabodurova committed:
fixing checkstyle issues
